### PR TITLE
[5.0] Allow OS color scheme override

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue #42134 .

### Summary of Changes

- introduce a selector for dark/light or OS controlled colour scheme
- Reflect the state to the HTML element with a `data-forced-theme` and a `data-theme="light|dark"`

TODO:

Create a new _root.scss so that the theme follows the states of the above HTML element states. (ie like: https://github.com/dgrammatiko/muta/blob/b675f730671e3dcf43e2c1331c6f7d80fcc6cb2c/media_source/templates/administrator/muta/scss/_root.scss#L1-L3)

### Testing Instructions

Check that in the Atum template the HTML element reflects the state of the User settings

### Actual result BEFORE applying this Pull Request

Atum Template is always set to the OS Color Scheme

### Expected result AFTER applying this Pull Request

Atum Template can override the OS Color Scheme

Screenshot 2023-10-23 at 20 46 44

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by richard67** [comment]
Why 2 binary switches? Why not one ternary option?
- value 0 „follow OS“ (= default)
- value 1 „light mode“
- value 2 „dark mode“

**Comment by dgrammatiko** [comment]
> Why 2 binary switches? Why not one ternary option?

Not a particular reason, I can change the code to that. A minor thing is that BS supports more than dark/light themes so limiting the options and also making the addition of new themes harder might not be the best option (future wise).

**Comment by richard67** [comment]
Let’s see if more opinions drop in regarding what’s better from a user point of view.

Other question: Shall we close the issue as we have a PR, or shall we leave it open until the ToDo is done?

File: administrator/language/en-GB/com_users.ini [hunk 1]
```diff
@@ -6,6 +6,10 @@
 COM_USERS="Users"
 COM_USERS_A11Y_SETTINGS_FIELDSET_DESC="These settings apply per user and are only available in supported templates eg Atum."
 COM_USERS_A11Y_SETTINGS_FIELDSET_LABEL="Accessibility Settings"
+COM_USERS_A11Y_SETTINGS_FIELD_ADMIN_PREFERS_COLOR_SCHEME_DESC="Sets the colour scheme for the Administrator template"
+COM_USERS_A11Y_SETTINGS_FIELD_ADMIN_PREFERS_COLOR_SCHEME_LABEL="Administrator Colour Scheme"
+COM_USERS_A11Y_SETTINGS_FIELD_SITE_PREFERS_COLOR_SCHEME_DESC="Sets the colour scheme for the Administrator template"
```
Comment (richard67) [inline administrator/language/en-GB/com_users.ini]:
```suggestion
COM_USERS_A11Y_SETTINGS_FIELD_SITE_PREFERS_COLOR_SCHEME_DESC="Sets the colour scheme for the Site template"
```

**Comment by dgrammatiko** [comment]
> Shall we close the issue as we have a PR, or shall we leave it open until the ToDo is done?


Meh, keep it open so people have a place to express their legitimate concerns 😀

File: administrator/language/en-GB/com_users.ini [hunk 2]
```diff
@@ -350,6 +356,7 @@ COM_USERS_OPTION_RANGE_TODAY="today"
 COM_USERS_OPTION_SELECT_COMPONENT="- Select Component -"
 COM_USERS_OPTION_SELECT_LEVEL_END="- Select End Level -"
 COM_USERS_OPTION_SELECT_LEVEL_START="- Select Start Level -"
+COM_USERS_OS_CHOICE="OS Choise"
```
Comment (richard67) [inline administrator/language/en-GB/com_users.ini]:
```suggestion
COM_USERS_OS_CHOICE="OS Choice"
```

**Comment by richard67** [comment]
The description of this PR should be adapted to the change to one tristate (list) parameter.

**Comment by wilsonge** [comment]
We're going to need two different compilations of the css file to support this and bundle that. I really don't see the need. More people have moaned about this because they don't like the design which should be fixed in the other PR than because they actually want a dark mode browser with a bright site. I think we should focus efforts on making the styling of dark mode better than the all black thing I did (obviously I'm not the worlds best designer) than adding extra options and overhead.

**Comment by dgrammatiko** [comment]
> We're going to need two different compilations of the css file to support this and bundle that.


Nope, check the link in the description, there is only one cas file!

**Comment by HLeithner** [comment]
why don't we go the bootstrap why? 
Adding the switch functionality to `Joomla.themeSelector` javascript class similar what's used by bootstrap saving the manual switch in the browser local storage and allow to use Javascript Options be loaded from the use personal settings?

btw. I wouldn't separated front and backend, we already have to many switchs.

**Comment by dgrammatiko** [comment]
> why don't we go the bootstrap way?

The bootstrap way:
Screenshot 2023-10-23 at 22 41 15

Joomla5: 
Screenshot 2023-10-23 at 20 46 44 2

The difference is that the state is saved in the DB which (for the logged in users) doesn't need writing in the users HD (local storage as cookies are data written in the disk). The JS switch could save the state to either: local storage, session storage, indexDB, cookie which only session storage is on the RAM (so the other require permission). 
In short saving the state in the DB is one of the possible solutions but not the only one (ie: in Muta I'm using cookies).

> btw. I wouldn't separated front and backend, we already have to many switchs.

Can do

File: administrator/templates/atum/component.php [hunk 3]
```diff
@@ -47,10 +47,18 @@
 $this->addHeadLink(HTMLHelper::_('image', 'favicon.ico', '', [], true, 1), 'alternate icon', 'rel', ['type' => 'image/vnd.microsoft.icon']);
 $this->addHeadLink(HTMLHelper::_('image', 'joomla-favicon-pinned.svg', '', [], true, 1), 'mask-icon', 'rel', ['color' => '#000']);
 
+// Getting user accessibility settings
+$user            = $app->getIdentity();
+$userScheme      = $user->getParam('admin_prefers_color_scheme', '0');
+$forcedSchemeAtr = $forcedScheme ? 'data-forced-theme ' : '';
```
Comment (richard67) [inline administrator/templates/atum/component.php:53]: Where does `$forcedScheme` come from? Can it be the logic here needs a change after the change to tristate?

Comment (richard67) [reply administrator/templates/atum/component.php:53]: Hmm, I think the above line 53 just needs to be removed. Same for the other PHP files of course.

Comment (dgrammatiko) [reply administrator/templates/atum/component.php:53]: It supposed to be `forcedSchemeAtr = '';`

File: administrator/templates/atum/error_full.php [hunk 4]
```diff
@@ -62,14 +62,14 @@
...
-			--template-special-color: ' . $this->params->get('special-color', '#001B4C') . ';
-		}');
+            --hue: ' . $matches[1] . ';
+            --template-bg-light: ' . $this->params->get('bg-light', '#f0f4fb') . ';
+            --template-text-dark: ' . $this->params->get('text-dark', '#495057') . ';
+            --template-text-light: ' . $this->params->get('text-light', '#ffffff') . ';
+            --link-color: ' . $linkColor . ';
+            --link-color-rgb: ' . $r . ',' . $g . ',' . $b . ';
+            --template-special-color: ' . $this->params->get('special-color', '#001B4C') . ';
+        }');
```
Comment (richard67) [inline administrator/templates/atum/error_full.php:72]: I think the "code style" change here should be reverted. It's in the inline CSS not the PHP code, so I think the tabs are ok, also for consistency with the other template PHP files.

**Comment by bembelimen** [comment]
I like the switch, but I think we should not create our own wheel again and just using what we [already have in Joomla](https://github.com/twbs/bootstrap/blob/main/site/static/docs/5.3/assets/js/color-modes.js). Also only one attribute is needed but [we have to change this parameter](https://github.com/joomla/joomla-cms/blob/5.0-dev/build/media_source/templates/administrator/atum/scss/_variables.scss#L5) to e.g. "data" (or something else and fix the broken and wrongly added definitions).

You can test this by adding data-bs-theme="dark|light" attribute to the html tag in joomla backend. So please don't add another dark/light mode layer on top of [the existing one](https://getbootstrap.com/docs/5.3/customize/color-modes/#dark-mode).

**Comment by dgrammatiko** [comment]
> we should not create our own wheel

It's just a DB value for a given key, people/devs could use it or not. Ie if I want to roll out something that is not requiring permissions (due to EU law about asking before writing to users HD) then such a value is the only way (also it doesn't require JS, not tightly coupled to Bootstrap, etc).

Anyways I'm closing as it's not worthy our time

**Comment by simbus82** [comment]
> More people have moaned about this because they don't like the design which should be fixed in the other PR than because they actually want a dark mode browser with a bright site.

This is your personal thinking. Facts are different.

**Comment by wilsonge** [comment]
Actually it’s based off the direct feedback that we had in mattermost from a dozen people during the betas and RCs. What’s your facts based on?

**Comment by simbus82** [comment]
> Actually it’s based off the direct feedback that we had in mattermost from a dozen people during the betas and RCs. What’s your facts based on?

Some billions of people 😅
https://github.com/joomla/joomla-cms/issues/42134#issuecomment-1768387471
